### PR TITLE
Rewrite futures-core docs

### PR DESCRIPTION
**Pull request by aturon**
Note that module-level docs are mostly removed here, since the ones in the facade are more important.

We may ultimately want to use the recent rustdoc file inclusion features to factor out and share those docs, once the facade ones have been updated.

### CODE REVIEW

File: futures-core/src/executor.rs [hunk 1]
```diff
@@ -1,32 +1,56 @@
...
+//! Executors.
 
 if_std! {
     use std::boxed::Box;
     use Future;
 
-    /// TODO: dox
+    /// A task executor.
+    ///
+    /// A *task* is a `()`-producing future that runs at the top level, and will
```
Comment (cramertj) [inline futures-core/src/executor.rs]: Might be worth mentioning that Tasks are also the unit at which wakeups occur-- I get questions about this a lot.

File: futures-core/src/future/mod.rs [hunk 2]
```diff
@@ -12,158 +11,114 @@ mod option;
...
-/// In addition to the documentation here you can also find more information
-/// about futures [online] at [https://tokio.rs](https://tokio.rs)
-///
-/// [online]: https://tokio.rs/docs/getting-started/futures/
-///
-/// Futures are used to provide a sentinel through which a value can be
-/// referenced. They crucially allow chaining and composing operations through
-/// consumption which allows expressing entire trees of computation as one
-/// sentinel value.
+/// In a way, a future is like a `Result` value that may not have finished
```
Comment (cramertj) [inline futures-core/src/future/mod.rs]: s/"In a way"//

File: futures-core/src/future/mod.rs [hunk 3]
```diff
@@ -12,158 +11,114 @@ mod option;
...
+/// computing yet. This kind of "asynchronous value" makes it possible for a
+/// thread to continue doing useful work while it waits for the value to become
+/// available.
 ///
 /// The ergonomics and implementation of the `Future` trait are very similar to
-/// the `Iterator` trait in that there is just one methods you need
-/// to implement, but you get a whole lot of others for free as a result.
+/// the `Iterator` trait in that there is just one methods you need to
```
Comment (cramertj) [inline futures-core/src/future/mod.rs]: *method

File: futures-core/src/future/mod.rs [hunk 4]
```diff
@@ -12,158 +11,114 @@ mod option;
...
-/// A key aspect of this method is that if the value is not yet available the
-/// current task is scheduled to receive a notification when it's later ready to
-/// be made available. This follows what's typically known as a "readiness" or
-/// "pull" model where values are pulled out of futures on demand, and
-/// otherwise a task is notified when a value might be ready to get pulled out.
-///
-/// The `poll` method is not intended to be called in general, but rather is
-/// typically called in the context of a "task" which drives a future to
-/// completion. For more information on this see the `task` module.
+/// The core method of future, `poll`, *attempt* to resolve the future into a
```
Comment (cramertj) [inline futures-core/src/future/mod.rs]: attempts

File: futures-core/src/future/mod.rs [hunk 5]
```diff
@@ -12,158 +11,114 @@ mod option;
...
+/// final value. This method does not block if the value is not ready. Instead,
+/// the current task is scheduled to be woken up when it's possible to make
```
Comment (cramertj) [inline futures-core/src/future/mod.rs]: "the current task"-- might it be clearer to refer to `cx.waker()` directly?

File: futures-core/src/future/mod.rs [hunk 6]
```diff
@@ -12,158 +11,114 @@ mod option;
...
+/// further progress by `poll`ing again. In this way, the futures task system
```
Comment (cramertj) [inline futures-core/src/future/mod.rs]: Not sure this last sentence is necessary

File: futures-core/src/future/mod.rs [hunk 7]
```diff
@@ -12,158 +11,114 @@ mod option;
...
-/// or those on `Option` and `Result`. Like with iterators, the combinators are
-/// zero-cost and don't impose any extra layers of indirection you wouldn't
-/// otherwise have to write down.
-///
-/// More information about combinators can be found [on tokio.rs].
-///
-/// [on tokio.rs]: https://tokio.rs/docs/going-deeper-futures/futures-mechanics/
+/// scheduling a number of callbacks. As with iterators, the combinators are
+/// zero-cost: they compile away. You can find the combinators in the
+/// [future-util](https://docs.rs/futures-util) crate.
```
Comment (cramertj) [inline futures-core/src/future/mod.rs]: Link to the `FutureExt` trait, specifically

File: futures-core/src/future/mod.rs [hunk 8]
```diff
@@ -12,158 +11,114 @@ mod option;
...
-    /// not work properly otherwise.
-    ///
-    /// When a future is not ready yet, the `Async::Pending` value will be
-    /// returned. In this situation the future will *also* register interest of
-    /// the current task in the value being produced. This is done by calling
-    /// `task::park` to retrieve a handle to the current `Task`. When the future
-    /// is then ready to make progress (e.g. it should be `poll`ed again) the
-    /// `unpark` method is called on the `Task`.
+    /// Attempt to resolve the future to a final value, registering
+    /// the current task for wakeup if the value is not yet available.
```
Comment (cramertj) [inline futures-core/src/future/mod.rs]: ditto about "current task"-- it seems like `cx.waker()` should be mentioned directly.

File: futures-core/src/stream/mod.rs [hunk 9]
```diff
@@ -1,93 +1,67 @@
...
-/// received from the same stream (it's valid to keep polling).
-///
-/// This property of streams, however, is [being considered] for change in 0.2
-/// where an error on a stream is similar to `None`, it terminates the stream
-/// entirely. If one of these use cases suits you perfectly and not the other,
-/// please feel welcome to comment on [the issue][being considered]!
-///
-/// [being considered]: https://github.com/alexcrichton/futures-rs/issues/206
+/// Streams, like futures, also bake in errors through an associated `Error`
+/// type.  an error on a stream **does not terminate the stream**. That is,
```
Comment (cramertj) [inline futures-core/src/stream/mod.rs]: s/a/A/

File: futures-core/src/stream/mod.rs [hunk 10]
```diff
@@ -1,93 +1,67 @@
...
 
     /// Attempt to pull out the next value of this stream, returning `None` if
     /// the stream is finished.
     ///
-    /// This method, like `Future::poll`, is the sole method of pulling out a
-    /// value from a stream. This method must also be run within the context of
-    /// a task typically and implementors of this trait must ensure that
-    /// implementations of this method do not block, as it may cause consumers
-    /// to behave badly.
+    /// This method, like [`Future::poll`](::future::Future::poll),
```
Comment (cramertj) [inline futures-core/src/stream/mod.rs]: Did you mean to have something after this comma?

File: futures-core/src/task/mod.rs [hunk 11]
```diff
@@ -1,31 +1,7 @@
-//! Tasks used to drive a future computation
+//! Task notification.
```
Comment (cramertj) [inline futures-core/src/task/mod.rs]: This whole file is awesome! I'm so excited to be able to point people to these docs :heart:

**Comment by aturon** [comment]
Pushed up fixes and rebased over the `poll_next` change.

File: futures-core/src/stream/mod.rs [hunk 12]
```diff
@@ -34,7 +34,10 @@ pub trait Stream {
     /// Attempt to pull out the next value of this stream, returning `None` if
     /// the stream is finished.
     ///
-    /// This method, like [`Future::poll`](::future::Future::poll),
+    /// This method, like [`Future::poll`](::future::Future::poll), makes
```
Comment (cramertj) [inline futures-core/src/stream/mod.rs]: This sentence feels a bit awkward to me. The docs for `Future::poll` say "Attempt to resolve the future to a final value, registering the current task for wakeup if the value is not yet available." Perhaps we could do something similar here:

"Attempt to resolve the next value in the stream, registering the current task for a wakeup if the value is not yet available."

**Comment by cramertj** [comment]
LGTM pending [the change to the `Stream::poll_next` docs](https://github.com/rust-lang-nursery/futures-rs/pull/782#discussion_r169725620).

**Pull request merged**